From bf4610c947c7dc372c4078f363d2dff6ae0703a8 Mon Sep 17 00:00:00 2001
From: iglocska <andras.iklody@gmail.com>
Date: Mon, 13 Jul 2020 12:19:11 +0200
Subject: [PATCH] fix: [security] setting a favourite homepage was not CSRF
 protected
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

- a user could be lured into setting a MISP home-page outside of the MISP baseurl
- switched the endpoint to be CSRF protection enabled

- as discovered by Mislav Božičević <mislav.bozicevic@nn.cz>
---
 app/Controller/AppController.php  |  2 +-
 app/View/Elements/global_menu.ctp |  5 +++--
 app/View/Layouts/default.ctp      |  1 +
 app/webroot/js/misp.js            | 22 +++++++++++++++-------
 4 files changed, 20 insertions(+), 10 deletions(-)

diff --git a/app/Controller/AppController.php b/app/Controller/AppController.php
index 958eae3b1d..e8f6003ff6 100755
--- a/app/Controller/AppController.php
+++ b/app/Controller/AppController.php
@@ -46,7 +46,7 @@ class AppController extends Controller
 
     public $helpers = array('Utility', 'OrgImg', 'FontAwesome', 'UserName', 'DataPathCollector');
 
-    private $__queryVersion = '106';
+    private $__queryVersion = '107';
     public $pyMispVersion = '2.4.128';
     public $phpmin = '7.2';
     public $phprec = '7.4';
diff --git a/app/View/Elements/global_menu.ctp b/app/View/Elements/global_menu.ctp
index 6bb141a74c..5cfa1ebb55 100755
--- a/app/View/Elements/global_menu.ctp
+++ b/app/View/Elements/global_menu.ctp
@@ -413,8 +413,9 @@
                 'type' => 'root',
                 'url' => '#',
                 'html' => sprintf(
-                    '<span class="fas fa-star %s" id="setHomePage" title="Set the current page as your home page in MISP"></span>',
-                    (!empty($homepage['path']) && $homepage['path'] === $this->here) ? 'orange' : ''
+                    '<span class="fas fa-star %s" id="setHomePage" title="Set the current page as your home page in MISP" data-current-page="%s"></span>',
+                    (!empty($homepage['path']) && $homepage['path'] === $this->here) ? 'orange' : '',
+                    $this->here
                 )
             ),
             array(
diff --git a/app/View/Layouts/default.ctp b/app/View/Layouts/default.ctp
index 9f5b7d1694..57f391f9a5 100644
--- a/app/View/Layouts/default.ctp
+++ b/app/View/Layouts/default.ctp
@@ -88,6 +88,7 @@
     <div id = "ajax_fail_container" class="ajax_container">
         <div id="ajax_fail" class="ajax_result ajax_fail"></div>
     </div>
+    <div id = "ajax_hidden_container" class="hidden"></div>
     <div class="loading">
         <div class="spinner"></div>
         <div class="loadingText"><?php echo __('Loading');?></div>
diff --git a/app/webroot/js/misp.js b/app/webroot/js/misp.js
index ffd2be285f..96949228cd 100644
--- a/app/webroot/js/misp.js
+++ b/app/webroot/js/misp.js
@@ -5022,15 +5022,23 @@ function resetDashboardGrid(grid) {
 
 function setHomePage() {
     $.ajax({
-        type: 'POST',
+        type: 'GET',
         url: baseurl + '/userSettings/setHomePage',
-        data: {
-            path: window.location.pathname
-        },
         success:function (data, textStatus) {
-            showMessage('success', 'Homepage set.');
-            $('#setHomePage').addClass('orange');
-        },
+            $('#ajax_hidden_container').html(data);
+            var currentPage = $('#setHomePage').data('current-page');
+            $('#UserSettingPath').val(currentPage);
+            $.ajax({
+                type: 'POST',
+                url: baseurl + '/userSettings/setHomePage',
+                data: $('#UserSettingSetHomePageForm').serialize(),
+                success:function (data, textStatus) {
+                    showMessage('success', 'Homepage set.');
+                    $('#setHomePage').addClass('orange');
+                },
+            });
+
+        }
     });
 }
 
